### ISSUE TITLE
Georeferencing hyperspectral data

### ISSUE THREAD

**Issue by dlebauer** [terraref/computing-pipeline#144]
Convert hyperspectral data cubes from pixel space to geographic space (x, y).

@czender needs script or mathematical formula(e) that takes meta-data and creates a coordinate system for hyperspectral data.

@solmazhajmohammadi has been working on taking meta-data and specifies geospatial coordinates in the gantry physical coordinate system.
- y is defined as 1mm (exactly?) per pixel.
- x depends on height of scanner

@smarshall-bmr still waiting on updated gantry positions.

**Comment by solmazhajmohammadi** [comment]
y is defined as  0.9853 mm per pixel. 
Exact number is 0.98526434004512529576754637665 mm

**Comment by czender** [comment]
@FlyingWithJerome please add bounding box coordinates (i.e., xy coordinates of the four corners of the image, upper-left, upper-right, lower-left, lower right) to the hyperspectral output file using whatever formulae @solmazhajmohammadi  recommends. This will be a good start. Adding coordinates for each pixel (i.e., x(x,y), and y(x,y)) might be overkill because it would triple the filesize, so let's just add the bounding box for now.

**Comment by czender** [comment]
@FlyingWithJerome @solmazhajmohammadi I added x and y offset coordinates to hyperspectral output file. The coordinates are placeholders for real values. Right now x=i*0.001 [m] and y=j*0.0009853 [m], as per above. I made-up x completely since I'm not sure how to compute it. We need to change these from offsets-from-origin to absolute coordinates.

**Comment by max-zilla** [comment]
@FlyingWithJerome here is the script @caicai89- has worked on:
https://github.com/terraref/computing-pipeline/blob/geostreams_extractor/scripts/geospatial/sensorposition.py

This pulls geospatial info from metadata and converts it into geometries describing the footprint of the file, and adds it to Clowder's PostGIS database.

This would be applied to any files with geospatial info in the metadata.json files.

**Comment by FlyingWithJerome** [comment]
@max-zilla Yes, Yaping's script does the similar job as we did; we need to convert the image coordinates to field coordinate and seems Yaping's script had already done so. But we still need each pixel in the graph to be traceable in absolute geographic position.

@dlebauer Should we continue developing the pixel2geogrphic, or it will be assigned to Yaping and we use the interface Yaping's script is going to provide?

**Comment by dlebauer** [comment]
@FlyingWithJerome Yaping's extractor should take the raw metadata and convert to a bounding box. 

He wasn't going to map geographic coordinates to pixels, but I would want to make sure that you are not duplicating effort with @solmazhajmohammadi, @ZongyangLi, or @pless. That is, if the same algorithm can be used for data from multiple sensors, please coordinate efforts and factor out the steps that can be generalized - even if it is just a set of parameters.

There are many reasons for this. Obviously, we want to use our time efficiently. Having three pieces of code that do the same thing is harder to understand and maintain than having one. On the other hand, it isn't always clear what should be generalized and it can be useful to come up with multiple solutions / drafts to the same problem before deciding whether and how to generalize. In short, please use your judgement!

**Comment by FlyingWithJerome** [comment]
@solmazhajmohammadi 
Can I have the X realworld length per pixel?

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome 
Imaging in X direction is similar to the normal camera. We can do the measurement for the lens distortion to get the exact length per pixel during the final test. But just for having a bounding box around the data, you can find the X length from the FOV and the distance from the camera. Focal length is small compare to the the working distance, so I ignored it. (25mm << 2m)

![eq](https://cloud.githubusercontent.com/assets/17046539/18443317/01fe317a-78da-11e6-9561-781737a868dd.jpg)

HFOV(Horizontal field of view) is length of X in real world. I think AFOV is 21 degree for SWIR camera and 44.6 degree for VNIR. WD is working distance (Z) which is 2 meter above the canopy. So at the end you will have the HFOV for 384 pixel in SWIR and 1600 pixels for the VNIR. 
Let me know if you need anything else.

**Comment by rachelshekar** [comment]
@FlyingWithJerome : can you update this issue?

**Comment by FlyingWithJerome** [comment]
@rachelshekar  Yes, we basically had the x and y coordinates for the pictures in hyperspectral outputs in both latitude longitude and xy forms, along with the Google map view for the location of the certain picture.

**Comment by rachelshekar** [comment]
@FlyingWithJerome - can this be closed?

**Comment by FlyingWithJerome** [comment]
Yes, I will close it

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome the current scans are done in ~50ms frame rate and 0.04m/s speed  (which is twice faster than requirement for external trigger signal). 
Meaning: just one out of two row scan is written into the hypercube. 

In this case(Frame rate> 25ms and scan speed = 0.04m/s) the data is downsampled by sampling every 2 element. you need to consider this to get the correct georefrenced hyeprcube.

**Comment by czender** [comment]
@FlyingWithJerome this means, if I understand correctly, that the y dimension in the images needs a more sophisticated formula for georeferencing. The current formula is described in the "y@algorithm" attribute:  

> Based on https://github.com/terraref/computing-pipeline/issues/144. The y-dimension spans 0.9853 mm per scanline/frame (exact number is 0.98526434004512529576754637665 mm). The datestamp for each frame is stored in the *_frameindex.txt file which is archived in the time variable.

We need to change this formula as @solmazhajmohammadi describes, though only under the conditions she describes. Please change the formula and update the attribute accordingly.

**Comment by remotesensinglab** [comment]
I noticed this issues was closed. But would like to know if it was resolved? Can we get georeferenced hyperspectral cubes stitched over the entire field?

**Comment by solmazhajmohammadi** [comment]
@remotesensinglab I dont think that we have stitched hyperspectral data, just because of the file size. But Charlie and Jerome are georefrencing the hyperspectral data
@czender @FlyingWithJerome 
Please note that this condition may change, as we are planning to come up with a solution to speed up the hyperspectral scan. Current scan speed is not matching with triggering signal and randomly sampling the data(as I described above). I'll update the issue, if any further changes are necessary.

**Comment by rachelshekar** [comment]
@solmazhajmohammadi - we need a calculation to convert output file to new speeds.

**Comment by dlebauer** [comment]
Would using position + velocity * time in the metadata provide sufficient accuracy?

**Comment by FlyingWithJerome** [comment]
@dlebauer @czender @solmazhajmohammadi 
I'm thinking position + velocity * time would be the basic of the new formula, but Solmaz also mentioned the frame rate is now larger than 25ms which causes downsampling. Probably we need more adjustments to the basic formula (twice the y scanline/frame?) as @czender mentioned above.

**Comment by czender** [comment]
@FlyingWithJerome please add arrays latitude(x) and longitude(y) using either pixel-based or velocity-based formulae, whichever you think is better, and we'll see how well it works. Neglect Earth's curvature, assume it's planar geometry. We should at some point discuss re-doing the coordinate system so it is based on a lower left (rather than lower right) corner reference and has longitude(x) and latitude(y), but that can wait.

**Comment by FlyingWithJerome** [comment]
@czender 
Will do. Planar latitude and longitude will probably be on in this weekend.

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome **position + velocity * time** is when system is running in free-run mode.
 In external triggering mode, system is triggered in every 0.9852 mm(in default speed = 0.02m/s). 
If the system is running twice of this speed with more than 25ms frame rate, the data is downsampled(explained above).

**Comment by FlyingWithJerome** [comment]
@solmazhajmohammadi By "downsampling" in sampling every one out of two lines, does it mean that one line will be thrown away, or two lines will be merged together and sampled them together?

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome It is a random process and the line will be thrown away(most likely first line will be kept)

**Comment by FlyingWithJerome** [comment]
@solmazhajmohammadi Thank you. If I understand you correctly, we will lose spatial information (because we scan one line, throw away one line, scan one line, throw away one line ....). And the pixel size won' change, because the line size remains the same

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome yes!

**Comment by FlyingWithJerome** [comment]
@solmazhajmohammadi 
Is there a flag in the metadata JSON so that our program can recognize that the picture is downsampled? By reading the "current setting speed" in the "sensor_variable_metadata" field, or speed [xyz] in the "gantry_system_variable_metadata?"

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome
 In JSON files, you can check "scanSpeedInMPerS [m/s]" and  "current setting frameperiod"  in the metadata. 
If the scan speed("scanSpeedInMPerS [m/s]") is "0.04" and frame period ("current setting frameperiod" )is "50"(more than 25), then it means that data is downsampled.

**Comment by czender** [comment]
@solmazhajmohammadi @FlyingWithJerome we need a "formula" that unambiguously determines when data is downsampled. please confirm that the formula is:
```
if((scanSpeedInMPerS == 0.04) and (current setting frameperiod > 25)) then
  downsampled == True;
else
  downsampled == False;
endif
```
If this is incorrect in any way, please provide a more accurate formula...

**Comment by max-zilla** [comment]
scan speed can be determined by the file size, we won't find that in the metadata if I understand @solmazhajmohammadi correctly.

**Comment by solmazhajmohammadi** [comment]
@FlyingWithJerome 
```
if(scanSpeedInMPerS is recrded in metadata) then
        if((scanSpeedInMPerS == 0.04) and (current setting frameperiod > 25)) 
              downsampled == True;
        elseif((scanSpeedInMPerS == 0.02) and (current setting frameperiod > 25))
              downsampled == False;
        else
              Error; 
        endif
else
        if((lines == 10000± 2500) and ("Position y [m]" > 18 or "Position y [m]" <2))
               downsampled == True;
        else
              downsampled == False;
        endif
endif
```

"lines" are specified in the header files(XXXX_raw.hdr). 
The Error part is to make sure that the system is not running in any random speed. In this case the downsampling process would be completely random. 
The second part of the code(repeated below) can be used for the early stage dataset where metadata doesnt have "scanSpeedInMPerS" key.
```
if((lines == 10000± 2500) and ("Position y [m]" > 18 or "Position y [m]" <2))
          downsampled == True;
 else
         downsampled == False;
endif
```
```("Position y [m]" > 18 or "Position y [m]" <2)``` is to make sure that we are looking at full-row scans. This will cover almost all the cases, however there is a rare case that is not detectable by the above formula. Anyway this rare case is not gonna be a full row scan, so we are safe.

**Comment by FlyingWithJerome** [comment]
@solmazhajmohammadi @czender 
This makes sense to me. I read through some of the metadata and could not find the "scanSpeedInMPerS" variable. Will implement the new downsampled georeference algorithm from here.

**Comment by czender** [comment]
Thanks @solmazhajmohammadi. Hopefully this will fix the funny aspect ratio of downsampled images. If not, we've more work to do...

**Comment by solmazhajmohammadi** [comment]
ScanspeedinMPerS was added recently with new software update. I think the data after October should have the scan speed in metadata. For the rest of the data hopefully the formula should fix the issue. Please Let me know if you have any issue.

Sent from my iPhone

On Mar 9, 2017, at 3:36 PM, Charlie Zender <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks @solmazhajmohammadi<https://github.com/solmazhajmohammadi>. Hopefully this will fix the funny aspect ratio of downsampled images. If not, we've more work to do...

**Comment by max-zilla** [comment]
The downsampling has been fixed and terrautils has standard routines we will use for georeferencing.

**Comment by solmazhajmohammadi** [comment]
@smarshall-bmr 
Based on the recent update on SWIR, I remember that you have observed that each SWIR line is about 2mm(twice the VNIR line).
Can you please confirm this and update this issue for @czender to update the pipeline.
